### ISSUE TITLE
Document swapfile and hibernation

### ISSUE THREAD

**Issue by kdave** [kdave/btrfs-progs#544]
https://lore.kernel.org/all/31660c315eeba4c461b6006b6d798355696d2155.camel@scientia.org/

In short: hibernation needs physical device offsets, the mapping needs to be done from btrfs file logical offset -> btrfs device offset -> block device offset. There's a tool to provide the offsets, but at least this should be documented.

**Comment by kdave** [comment]
https://github.com/osandov/osandov-linux/blob/master/scripts/btrfs_map_physical.c

**Comment by calestyo** [comment]
btw: this is a duplicate of my earlier #533.

And I don't think it's enough to just provide a link to Omar's tool... at least not for many end-users who'll even have troubles compiling it.
Not to talk, that it wouldn't get any updates (including any possible security updates), as it wouldn't be part of the distributions.

The tool, should be simply included in `btrfs-progs`. :-)

**Comment by kdave** [comment]
Yes of course, I dropped the link so I can find it once I start working on it.

**Comment by kdave** [comment]
Pushed to branch `map-swapfile`, a new subcommand for inspect that verifies the extents and prints the first offset. It's almost finished but the output may change according to feedback.

**Comment by kdave** [comment]
Actually the value needs to be divided by page size in order to be used for `resume_offset`.

**Comment by calestyo** [comment]
I guess it make sense for btrfsprogs to provide that as an independent command (i.e. `map_physical`) for general use, but wouldn't that work for **any** files and not just swapfiles... so shouldn't the command be named more general?

However, if it's just for swapfiles **and** just for getting the resume_offset... it should IMO also somehow reflect that perhaps in the name? And also really give out the value one has to enter in `resume_offset` (i.e. already divided by the page size).<br>
In that case, there should be some documentation on how users find out their page size in order to get to the `resume_offset` value.

btw: I think the resume system's interface there is not the best for end-users… it would be much better if there was simply a `resume_file` and the kernel could find out the rest on its own.

**Comment by kdave** [comment]
I think a specific command for the swapfile is convenient as it also checks that all the swapfile requirements are OK, ie. no holes, compression and stored on one device on single type chunks. This is inherited from btrfs_map_physical, only skips printing the extents. Another command for listing extents will be also added, basically what the map_physical tool does.

The name can be changed, the `map-` prefix may be misleading but remotely it does some mapping and I'd really like to keep `swapfile` in the name because it's very specific to that. The printed value can be tuned by parameters, primary would be the `resume_offset` so it can be directly used for `/sys/power/resume_offset`. Page size can be read from `getconf` (same command exists).

Agreed that the user interface for resume is not the best, I'd think it's also fragile and inflexible. Writing to the swapfile could relocate the first block so it could get out of sync with the kernel command line parameters that can't be changed on a running system. Ie. always use the sysfs file for resume_offset. And it also needs the name of the physical device, that can potentially get a different name after boot (maybe not, some buses are iterated each time, some are restored from memory).

**Comment by calestyo** [comment]
> I think a specific command for the swapfile is convenient as it also checks that all the swapfile requirements are OK, ie. no holes, compression and stored on one device on single type chunks. This is inherited from btrfs_map_physical, only skips printing the extents. Another command for listing extents will be also added, basically what the map_physical tool does.

Ah ok, that's nice then.

> The name can be changed, the map- prefix may be misleading but remotely it does some mapping and I'd really like to keep swapfile in the name because it's very specific to that.

It may be confusing when the one `map-` command would give the offset in bytes, and the other one for the swapfile in page sizes.

So perhaps directly something something with `resume-file` like `resume-file-offset`? It's anyway clear that the resume file would be a swap file.<br>
However, that name wouldn't reflect that you do more checks.

> Agreed that the user interface for resume is not the best, I'd think it's also fragile and inflexible. Writing to the swapfile could relocate the first block so it could get out of sync with the kernel command line parameters that can't be changed on a running system. Ie. always use the sysfs file for resume_offset. And it also needs the name of the physical device, that can potentially get a different name after boot (maybe not, some buses are iterated each time, some are restored from memory).

Any chances that simply that interface is improved?

**Comment by kdave** [comment]
It's in devel now (experimental build) and for hibernation can be used like:

```
# btrfs inspect map-swapfile -r /path/to/swapfile
198122980
```

And full output is
```
# btrfs inspect map-swapfile /path/to/swapfile
Physical start: 811511726080
Resume offset:     198122980
```

**Comment by calestyo** [comment]
Thanks for your work on this :-)
I made a few further comments… but these are just cosmetic things.
